lib/Sema/TypeCheckType.cpp

```diff
@@ -2414,20 +2414,21 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
       if (!attrs.has(i))
         continue;
 
-      auto diag = diagnoseInvalid(repr, attrs.getLoc(i),
-                                  diag::attribute_requires_function_type,
-                                  TypeAttributes::getAttrName(i));
-
-      // If we see @escaping among the attributes on this type, because it isn't
-      // a function type, we'll remove it.
-      if (i == TAK_escaping) {
-        diag.fixItRemove(getTypeAttrRangeWithAt(Context,
-                                                attrs.getLoc(TAK_escaping)));
-        // Specialize the diagnostic for Optionals.
-        if (ty->getOptionalObjectType()) {
-          diag.flush();
-          diagnoseInvalid(repr, repr->getLoc(),
-                          diag::escaping_optional_type_argument);
+      if (i == TAK_escaping && ty->getOptionalObjectType()) {
```

Comment:
Unfortunately checking for `ty->getOptionalObjectType()` is not enough, to produce `escaping_optional_type_argument` we have to make sure that object type is indeed a function type (`AnyFunctionType` to be precise).  I think it makes sense to try and unify this if/else into a single `if` for `i == TAK_escaping` and have some additional logic to pick which message to produce inside it because both cases share logic to emit a fix-it.

Reply:
So, I tried but couldn't unify if/else into a single `if` for `i == TAK_escaping` 😕
We can't reassign InFlightDiagnostic so I couldn't reuse one `fixItRemove`

Reply:
You could define a lambda for the correct `InFlightDiagnostic` outside the loop and call it instead of `diagnoseInvalid`. As a bonus, we get to adjust the flow to only check against an optional when `i == TAK_escaping`.

Reply:
I didn't catch the idea with lambda 😢

If we declare lambda outside it will anyway produce `InFlightDiagnostic` that we can't reassign so we'll still have to use double-check `i == TAK_escaping` and double-use `fixItRemove`
Am I wrong?

Reply:
The lambda emits the correct diagnostic and returns an `InFlightDiagnostic` you can then attach stuff to; no need to reassign anything:
```c++
const auto diagnoseInvalidAttr = [&](TypeAttrKind kind) {
  if (kind == TAK_escaping) {
    Type optionalObjectType = ty->getOptionalObjectType();
    if (optionalObjectType && optionalObjectType->is<AnyFunctionType>())
      return diagnoseInvalid(...);
  }
  return diagnoseInvalid(...);
};
```

Reply:
Agree with @AnthonyLatsis, You can even pass the attr so you don't have to define the lambda inside the loop :)

```cpp
    const auto diagnoseInvalidAttr = [&](Type attributeType, TypeAttrKind attrKind) {
      if (attributeType &&
          attributeType->is<AnyFunctionType>()) {
        return diagnoseInvalid(repr, attrs.getLoc(attrKind), ...);
      } else {
        return diagnoseInvalid(...);
      }
    };
    
    for (auto i : FunctionAttrs) {
      if (!attrs.has(i))
        continue;
      Type optionalObjectType = ty->getOptionalObjectType();
      if (i == TAK_escaping) {
        diagnoseInvalidAttr(optionalObjectType, i)
          .fixItRemove(
                getTypeAttrRangeWithAt(Context, attrs.getLoc(TAK_escaping)));
     
      }
      attrs.clearAttribute(i);
    }
```

Reply:
Oh my gosh, thanks a lot, now it looks way better

Reply:
@LucianoPAlmeida I'm not sure that it will work when `i != TAK_escaping`, will it?

Reply:
Ahh, sure ... this was just a sketch real quick to show a possible approach :) 
I think you right, it has to work for other function attrs the only applies to function types e.g. `@autoclosure`